Reflection_Engine/Compute/ReflectNonBHoMTypes.cs

```diff
@@ -0,0 +1,87 @@
+ * Copyright (c) 2015 - 2021, the respective contributors. All rights reserved.
+ *
+ * Each contributor holds copyright over their respective contributions.
+ * The project versioning (Git) records all such contribution source information.
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.oM.Base;
+using BH.oM.Reflection.Attributes;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Reflection;
+using System.Text.RegularExpressions;
+
+namespace BH.Engine.Reflection
+{
+    public static partial class Compute
+    {
+        /***************************************************/
+        /****               Public methods              ****/
+        /***************************************************/
+
+        [Description("Extracts all types from an already loaded non-BHoM assembly and adds them to the Global collection of all reflected types.")]
+        [Input("assembly", "Non-BHoM assembly with the types to be reflected.")]
+        public static void ReflectNonBHoMTypes(Assembly assembly)
+        {
+            if (assembly == null || !Global.AllAssemblies.ContainsKey(assembly.FullName) || assembly.ReflectionOnly)
+            {
+                RecordError($"{nameof(ReflectNonBHoMTypes)} method works only with assemblies loaded to BHoM.");
+                return;
+            }
+
+            if (m_NonBHoMTypesReflected.Contains(assembly.FullName) || assembly.IsBHoM())
+                return;
+
+            lock (Global.AssemblyReflectionLock)
+            {
+                foreach (Type type in assembly.GetTypes())
+                {
+                    if (type.Namespace != null && !type.IsAutoGenerated())
+                        Global.AllTypeList.Add(type);
+                }
+
+                m_NonBHoMTypesReflected.Add(assembly.FullName);
+            }
```

Comment:
Think the lock needs to hit before you check if the assembly has already been reflected. That is the critical part. So something like this could do it:

```suggestion
            lock (Global.AssemblyReflectionLock)
            {
                if (m_NonBHoMTypesReflected.Contains(assembly.FullName) || assembly.IsBHoM())
                    return;

                m_NonBHoMTypesReflected.Add(assembly.FullName);

                foreach (Type type in assembly.GetTypes())
                {
                    if (type.Namespace != null && !type.IsAutoGenerated())
                        Global.AllTypeList.Add(type);
                }
            }
```

Then, actually, the looping through the types in the assembly should be fine to keep outside the lock really, as the AllTypeList is a ConcurrentCollection, so could be changing it to the following really:

```suggestion
            lock (Global.AssemblyReflectionLock)
            {
                if (m_NonBHoMTypesReflected.Contains(assembly.FullName) || assembly.IsBHoM())
                    return;

                m_NonBHoMTypesReflected.Add(assembly.FullName);
            }
            
            foreach (Type type in assembly.GetTypes())
            {
                 if (type.Namespace != null && !type.IsAutoGenerated())
                   Global.AllTypeList.Add(type);
            }
```

But that is not how we have it in the other method, so might opt for option 1 to start with.

Might also be good to put this in a try-catch, as has been done with the other methods.

(Also, code suggestions made in the browser so indentation is probably a bit off 😄 )